mm2src/common/mm_metrics.rs

```diff
@@ -0,0 +1,814 @@
+use crate::executor::{spawn, Timer};
+use crate::log::{LogArc, LogWeak, Tag};
+use gstuff::Constructible;
+use hdrhistogram::Histogram;
+use itertools::Itertools;
+use metrics_core::{Key, Label, Drain, Observe, Observer, ScopedString, Builder};
+use metrics_runtime::{Receiver, observers::PrometheusBuilder};
+pub use metrics_runtime::Sink;
+use metrics_util::{parse_quantiles, Quantile};
+use serde_json::{self as json, Value as Json};
+use std::collections::HashMap;
+use std::fmt::Write as WriteFmt;
+use std::ops::Deref;
+use std::slice::Iter;
+use std::sync::{Arc, Weak};
+
+/// Increment counter if an MmArc is not dropped yet and metrics system is initialized already.
+#[macro_export]
+macro_rules! mm_counter {
+    ($metrics:expr, $name:expr, $value:expr) => {{
+        if let Some(mut sink) = $crate::mm_metrics::TrySink::try_sink(&$metrics) {
+            sink.increment_counter($name, $value);
+        }
+    }};
+
+    ($metrics:expr, $name:expr, $value:expr, $($labels:tt)*) => {{
+        use metrics::labels;
+        if let Some(mut sink) = $crate::mm_metrics::TrySink::try_sink(&$metrics) {
+            let labels = labels!( $($labels)* );
+            sink.increment_counter_with_labels($name, $value, labels);
+        }
+    }};
+}
+
+/// Update gauge if an MmArc is not dropped yet and metrics system is initialized already.
+#[macro_export]
+macro_rules! mm_gauge {
+    ($metrics:expr, $name:expr, $value:expr) => {{
+        if let Some(mut sink) = $crate::mm_metrics::TrySink::try_sink(&$metrics) {
+            sink.update_gauge($name, $value);
+        }
+    }};
+
+    ($metrics:expr, $name:expr, $value:expr, $($labels:tt)*) => {{
+        use metrics::labels;
+        if let Some(mut sink) = $crate::mm_metrics::TrySink::try_sink(&$metrics) {
+            let labels = labels!( $($labels)* );
+            sink.update_gauge_with_labels($name, $value, labels);
+        }
+    }};
+}
+
+/// Pass new timing value if an MmArc is not dropped yet and metrics system is initialized already.
+#[macro_export]
+macro_rules! mm_timing {
+    ($metrics:expr, $name:expr, $start:expr, $end:expr) => {{
+        if let Some(mut sink) = $crate::mm_metrics::TrySink::try_sink(&$metrics) {
+            sink.record_timing($name, $start, $end);
+        }
+    }};
+
+    ($metrics:expr, $name:expr, $start:expr, $end:expr, $($labels:tt)*) => {{
+        use metrics::labels;
+        if let Some(mut sink) = $crate::mm_metrics::TrySink::try_sink(&$metrics) {
+            let labels = labels!( $($labels)* );
+            sink.record_timing_with_labels($name, $start, $end, labels);
+        }
+    }};
+}
+
+#[cfg(feature = "native")]
+pub mod prometheus {
```

Comment:
Did you have a chance to setup Prometheus demo? I think users will be glad to see an example of pretty metrics visualization that becomes available with this PR.

Reply:
I prepared a docker-compose yaml to run Prometheus and Grafana services configured to work with the Market Maker. I guess to write this in the documentation.

Reply:
Perfect! Can't wait to see this 🙂